src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3WebtransportManager.cs

```diff
@@ -0,0 +1,156 @@
+            }
+
+            QuicStream clientWebtransportStream;
+            try
+            {
+                clientWebtransportStream = await _connection.OpenOutboundStreamAsync(type, cancellationToken).ConfigureAwait(false);
+                await clientWebtransportStream.WriteAsync(BuildWebtransportStreamClientFrame(type, sessionId), cancellationToken).ConfigureAwait(false);
+
+                return clientWebtransportStream;
+            }
+            catch (Exception ex)
+            {
+                throw ex;
+            }
+        }
+
+        private byte[] BuildWebtransportStreamClientFrame(QuicStreamType type, long sessionId)
+        {
+            Span<byte> buffer = stackalloc byte[2 + VariableLengthIntegerHelper.MaximumEncodedLength];
+            long streamType = type == QuicStreamType.Unidirectional ? (long)Http3StreamType.WebTransportUnidirectional : (long)Http3StreamType.WebTransportBidirectional;
+            int webtransportLength = VariableLengthIntegerHelper.WriteInteger(buffer.Slice(0), streamType);
+            int webtransportSessionLength = VariableLengthIntegerHelper.WriteInteger(buffer.Slice(webtransportLength), sessionId);
+            int payloadLength = webtransportLength + webtransportSessionLength; // includes the webtransport stream and the session id
+            Debug.Assert(payloadLength <= VariableLengthIntegerHelper.OneByteLimit);
+
+            return buffer.Slice(0, payloadLength).ToArray();
+        }
+
+        public void DeleteSession(long id)
+        {
+            if (_disposed == 1)
+            {
+                throw new ObjectDisposedException(nameof(Http3WebtransportManager));
+            }
+
+            _sessions.TryRemove(id, out _);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (Interlocked.Exchange(ref _disposed, 1) == 1)
+                return;
+            List<Http3WebtransportSession> toRemove = new List<Http3WebtransportSession>();
+            lock (_sessions)
+            {
+                foreach (KeyValuePair<long, Http3WebtransportSession> pair in _sessions)
+                {
+                    toRemove.Add(pair.Value);
+                }
+                _sessions.Clear();
+            }
+            foreach (Http3WebtransportSession session in toRemove)
+            {
+                await session.DisposeAsync().ConfigureAwait(false);
+            }
+
+        }
+        public void Dispose()
+        {
+            if (Interlocked.Exchange(ref _disposed, 1) == 1)
+                return;
+            lock (_sessions)
+            {
+                List<Http3WebtransportSession> toRemove = new List<Http3WebtransportSession>();
+                foreach (KeyValuePair<long, Http3WebtransportSession> pair in _sessions)
+                {
+                    toRemove.Add(pair.Value);
+                }
+                _sessions.Clear();
+                foreach (Http3WebtransportSession session in toRemove)
```

Comment:
You can do that outside of lock, because you've already cleaned them from the dictionary, and with a small edit proposed below, nothing should be added to it anymore.

The point of synchronizing Dispose and Add is to make sure that we either collected *everything* in Dispose (Add won the race, so the session was already added by this time and it will make its way into toRemove), or Add fails (Dispose won the race, so by the time Add enters the lock, _disposed flag is already set).
To ensure that, you should also do _disposed check *inside* the lock in Add.

Reply:
This one is no longer actual, but added a lock on the session removal from dispose